velox/dwio/parquet/reader/PageReader.cpp

```diff
@@ -75,53 +75,64 @@ void PageReader::seekToPage(int64_t row) {
   }
 }
 
+// The size of Parquet Page Header remains unknown until the Header is parsed
+// because it is in Thrift format. To read Parquet Page Header, we are using an
+// adaptive approach to guess the Page Header size.
+
+// TODO: Refactor the inputStreams to have a general interface where read() can
+// be applied on different chunks of memory.
 PageHeader PageReader::readPageHeader(int64_t remainingSize) {
   // Note that sizeof(PageHeader) may be longer than actually read
   std::shared_ptr<thrift::ThriftBufferedTransport> transport;
   std::unique_ptr<apache::thrift::protocol::TCompactProtocolT<
       thrift::ThriftBufferedTransport>>
       protocol;
-  char copy[sizeof(PageHeader)];
-  bool wasInBuffer = false;
-  if (bufferEnd_ == bufferStart_) {
-    const void* buffer;
-    int32_t size;
-    inputStream_->Next(&buffer, &size);
-    bufferStart_ = reinterpret_cast<const char*>(buffer);
-    bufferEnd_ = bufferStart_ + size;
-  }
-  if (bufferEnd_ - bufferStart_ >= sizeof(PageHeader)) {
-    wasInBuffer = true;
-    transport = std::make_shared<thrift::ThriftBufferedTransport>(
-        bufferStart_, sizeof(PageHeader));
-    protocol = std::make_unique<apache::thrift::protocol::TCompactProtocolT<
-        thrift::ThriftBufferedTransport>>(transport);
-  } else {
+
+  uint32_t pageHeaderBufferSize = kPageHeaderBufferSize_;
+  uint32_t oldPageHeaderBufferSize = 0;
+
+  BufferPtr copyBuffer =
+      AlignedBuffer::allocate<char>(pageHeaderBufferSize, &pool_);
+
+  uint32_t attempt = 0;
+  for (attempt = 0; attempt < kMaxAttempts_; attempt++) {
```

Comment:
We don't need to limit the number of attempts, just keep trying until all `remainingSize` is read.